test/Abstractions.Tests/TaskNameTests.cs

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.DurableTask;
+
+public class TaskNameTests
+{
+    [Fact]
+    public void Ctor_NullName_Throws()
+    {
+        Func<TaskName> act = () => new TaskName(null!);
+        act.Should().ThrowExactly<ArgumentNullException>().WithParameterName("name");
+    }
+
+    [Fact]
+    public void Ctor_EmptyName_Throws()
+    {
+        Func<TaskName> act = () => new TaskName(string.Empty);
+        act.Should().ThrowExactly<ArgumentException>().WithParameterName("name");
+    }
+
+    [Theory]
+    [InlineData(null)]
+    [InlineData("")]
+    public void Conversion_EqualToDefault(string? str)
+    {
+        TaskName name = str!;
+        name.Should().Be(default(TaskName));
```

Comment:
It seems a little inconsistent that you can create a null/empty `TaskName` using the implicit operator but not explicitly using the constructor.

Reply:
Fair - would you like to allow non-null/empty values in the constructor? Or to throw in the implicit operator?

Reply:
I'm thinking it makes sense to throw in the implicit operator. However, I can also see an argument against that which is that implicit operators shouldn't throw (I don't know if that guideline exists, but it does seem strange that this is possible). So maybe we should just allow empty values and put the validation burden on the consumers of `TaskName`?

Reply:
Yes, there is a guideline to never throw in an implicit operator. I went with the 2nd option but had to do some slightly odd stuff that goes against explicit nullability. I had to get a null name match up to the default struct - which means version must also be null. I think this is acceptable for now, but may get ugly when version can be explicitly set.